### PR TITLE
fix: clang compile

### PR DESCRIPTION
**Pull request by ganler**
**Description**: cannot compile ORT-cuda with clang. 
It says "hides overloaded virtual function" which is a common clang compile error (also see [here](https://stackoverflow.com/questions/18515183/c-overloaded-virtual-function-warning-by-clang)).

**Motivation and Context**
- Fix ORT-cuda compilation on clang (specifically, my platform: Manjaro; clang-13)

image

### CODE REVIEW

**Comment by yuslepukhin** [comment]
> class RandomBase : public CudaKernel {

I think the inheritance needs to be fixed here. Let's do everything we need in the base, but inherit from CudaKernel the actual Kernels.

---
Refers to: onnxruntime/core/providers/cuda/generator/random.h:29 in 21bf3a9. [](commit_id = 21bf3a9a8faf79aec2225511bc0cd3de74bea5a8, deletion_comment = False)

**Comment by ganler** [comment]
@yuslepukhin Thaks for the reply. Did you mean we should fix it in the `RandomBase` class?

**Comment by yuslepukhin** [comment]
> @yuslepukhin Thaks for the reply. Did you mean we should fix it in the `RandomBase` class?

Yes, would you mind? Perhaps, rename Compute() to ComputeInternal() or ComputeBase() or ComputeHelper(). Other names may include what it actually doing. Having it inherit from CudaKernel and overload virtual method is not `comme il faut`.

**Comment by ganler** [comment]
@yuslepukhin I just implemented what you said. But instead of using virtual functions for `ComputeBase`, as here it shall not be a dynamic calling interface, I use CRTP pattern to implement the "virtual function" **statically**.

**Review by iK1D — APPROVED** [review]

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 1]
```diff
@@ -26,6 +26,7 @@ RANDOM_COMPUTE_IMPL(RandomUniform)
 
 #undef RANDOM_COMPUTE_IMPL
 
+template <typename RandomType>
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.h:29]: Let's try avoid unnecessary templating. This class simply provides functionality that is common and does not depend on any types.

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 2]
```diff
@@ -43,20 +44,26 @@ class RandomBase : public CudaKernel {
     }
   }
 
+  Status ComputeBase(OpKernelContext* p_ctx, const TensorShape& shape, int dtype) const {
+    // CRTP: `ComputeBaseImpl` needs to be implemented.
+    return static_cast<RandomType&>(*this).ComputeBaseImpl(p_ctx, shape, dtype);
+  }
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.h:50]: This function is not needed

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 3]
```diff
@@ -26,6 +26,7 @@ RANDOM_COMPUTE_IMPL(RandomUniform)
 
 #undef RANDOM_COMPUTE_IMPL
 
+template <typename RandomType>
 class RandomBase : public CudaKernel {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.h:30]: >public CudaKernel [](http://example.com/codeflow?start=19&length=17)

Let's make the resulting classes inherit from this.

**Comment by yuslepukhin** [comment]
>   }

We can make it a separate utility function in the RandomBase and invoke when necessary in the derived classes.

---
Refers to: onnxruntime/core/providers/cuda/generator/random.h:82 in 265c49d. [](commit_id = 265c49d926da8e74edf7b0d8fed6ec3e89659671, deletion_comment = False)

**Comment by yuslepukhin** [comment]
> class RandomNormal final : public RandomNormalBase {

Can we inherit directly from RandomBase?

---
Refers to: onnxruntime/core/providers/cuda/generator/random.h:73 in 265c49d. [](commit_id = 265c49d926da8e74edf7b0d8fed6ec3e89659671, deletion_comment = False)

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 4]
```diff
@@ -43,20 +44,26 @@ class RandomBase : public CudaKernel {
...
       ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;  // optional and may be inferred
 };
 
-class RandomNormalBase : public RandomBase {
+class RandomNormalBase : public RandomBase<RandomNormalBase> {
  protected:
   RandomNormalBase(const OpKernelInfo& info) : RandomBase(info) {
     ORT_ENFORCE(info.GetAttr<float>("scale", &scale_).IsOK());
     ORT_ENFORCE(info.GetAttr<float>("mean", &mean_).IsOK());
   }
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.h:63]: Make it a helper in the RandomBase

**Comment by yuslepukhin** [comment]
> class RandomNormalLike final : public RandomNormalBase {

Let's make it a simpler, inherit from RandomBase

---
Refers to: onnxruntime/core/providers/cuda/generator/random.h:90 in 265c49d. [](commit_id = 265c49d926da8e74edf7b0d8fed6ec3e89659671, deletion_comment = False)

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 5]
```diff
@@ -26,6 +26,7 @@ RANDOM_COMPUTE_IMPL(RandomUniform)
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.h:26]: These are used only in the .cc file. Let's move it there. Implementation detail.

**Comment by yuslepukhin** [comment]
> @yuslepukhin I just implemented what you said. But instead of using virtual functions for `ComputeBase`, as here it shall not be a dynamic calling interface, I use CRTP pattern to implement the "virtual function" **statically**.

Yes, I see it. I feel it makes things even more complicated. And templates generate more code here. We are aiming for less binary size. 

Let's make resulting kernels inherit from CudaKernel and not the base.
Remove ComputeBase and make ComputeInternal() overrides call ComputeBaseImpl() directly.
Pass const CudaKernel* (essentially *this) as a parameter into ComputeBaseImpl() to make necessary kernel functionality available.

**Comment by ganler** [comment]
Thank you. We can discuss under your PR. https://github.com/microsoft/onnxruntime/pull/10160

**Pull request closed without merging**